### PR TITLE
Update backend [part 4]

### PR DESCRIPTION
**Pull request by r0mainK**
Take out unused methods in backend relating to the index, add `initialize_backend` method, create helper function in gcs_backend to avoid redundancy

### CODE REVIEW

**Comment by vmarkovtsev** [comment]
@r0mainK There are merge conflicts

**Comment by r0mainK** [comment]
@vmarkovtsev solved them, ready for review when you get the time

File: modelforge/registry.py [hunk 1]
```diff
@@ -24,19 +23,10 @@ def publish_model(args: argparse.Namespace, backend: StorageBackend, log: loggin
...
     path = os.path.abspath(args.model)
-    try:
-        model = GenericModel(source=path, dummy=True)
-    except ValueError as e:
-        log.critical('"model" must be a path: %s', e)
-        return 1
-    except Exception as e:
-        log.critical("Failed to load the model: %s: %s" % (type(e).__name__, e))
-        return 1
+    model = GenericModel(source=path, dummy=True)
```
Comment (vmarkovtsev) [inline modelforge/registry.py]: Why did you remove the error handlers? It is a top level function

File: modelforge/registry.py [hunk 2]
```diff
@@ -24,19 +23,10 @@ def publish_model(args: argparse.Namespace, backend: StorageBackend, log: loggin
...
     base_meta = model.meta
-    try:
-        model_url = backend.upload_model(path, base_meta, args.force)
-    except ModelExistsError:
-        return 1
+    model_url = backend.upload_model(path, base_meta, args.force)
```
Comment (vmarkovtsev) [inline modelforge/registry.py]: Same here

File: modelforge/gcs_backend.py [hunk 3]
```diff
@@ -75,31 +75,41 @@ def bucket_name(self):
...
             client = Client()
+        return client
+
+    def connect(self):
+        log = self._log
+        log.info("Connecting to the bucket...")
+        client = self.create_client()
         return client.lookup_bucket(self.bucket_name)
 
+    def initialize_backend(self, force):
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply modelforge/gcs_backend.py]: done

File: modelforge/gcs_backend.py [hunk 4]
```diff
@@ -75,31 +75,41 @@ def bucket_name(self):
...
     def upload_model(self, path, meta, force):
         bucket = self.connect()
         if bucket is None:
-            raise TransactionRequiredError
+            raise BackendRequiredError
         blob = bucket.blob("models/%s/%s.asdf" % (meta["model"], meta["uuid"]))
         if blob.exists() and not force:
             self._log.error("Model %s already exists, aborted.", meta["uuid"])
-            return 1
+            raise ExistingModelError
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply modelforge/gcs_backend.py]: done

File: modelforge/gcs_backend.py [hunk 5]
```diff
@@ -124,23 +134,14 @@ def request_with_progress(method, url, data=None, headers=None, **kwargs):
...
-        blob.upload_from_string(json.dumps(index, indent=4, sort_keys=True))
-        blob.make_public()
-
-    def _fetch(self, url, where, chunk_size=DEFAULT_CHUNK_SIZE):
-        self._log.info("Fetching %s...", url)
-        r = requests.get(url, stream=True)
-        if isinstance(where, str):
-            os.makedirs(os.path.dirname(where), exist_ok=True)
-            f = open(where, "wb")
+    def fetch_model(self, source: str, file: str, chunk_size: int =DEFAULT_CHUNK_SIZE) -> None:
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply modelforge/gcs_backend.py]: my bad, changing

[the remark this answers is not in the document]
Comment (r0mainK) [reply modelforge/gcs_backend.py]: also added typing to the rest of the file

File: modelforge/gcs_backend.py [hunk 6]
```diff
@@ -124,23 +134,14 @@ def request_with_progress(method, url, data=None, headers=None, **kwargs):
...
+        self._log.info("Fetching %s...", source)
+        r = requests.get(source, stream=True)
```
[the remark this answers is not in the document]
Comment (r0mainK) [reply modelforge/gcs_backend.py]: done

File: modelforge/registry.py (hunk 1 above)
Comment (r0mainK) [reply modelforge/registry.py]: here was my mistake, reverting

File: modelforge/registry.py (hunk 2 above)
Comment (r0mainK) [reply modelforge/registry.py]: here I don't understand why, since the error is a custom one that clearly explains whats happening - and logs in `upload_model` also provide information. Unless you want to sytematically wrap errors this way to return 1 ? In which case we should wrap the other top level functions too, no ?

Comment (r0mainK) [reply modelforge/registry.py]: anyway reverting

Comment (vmarkovtsev) [reply modelforge/registry.py]: Yep we should wrap everything, that was a dirty cutting off the edges

**Comment by r0mainK** [comment]
Added catches about everywhere, and moved the message to the logger in most cases. I put two TODO for when dulwich implements a porcelain error. Gonna add some test now, to get coverage a bit higher. But yeah ready for more review or merge after that. 
I'm also gonna reorder the methods and functions so they are more coherent.

File: modelforge/registry.py [hunk 7]
```diff
@@ -6,12 +6,56 @@
...
+    :param log: Logger supplied by supply_backend
+    :return: None
+    """
+    try:
+        backend.reset(args.force)
+    except ExistingBackendError:
+        return 1
+
+    log.info("Deleting content of index ...")
+    backend.index.initialize_index()
```
Comment (vmarkovtsev) [inline modelforge/registry.py]: Future commit: rename `initialize_index` to `reset`

File: modelforge/registry.py [hunk 8]
```diff
@@ -6,12 +6,56 @@
...
+    try:
+        backend.index.upload_index("initilialize", {})
```
Comment (vmarkovtsev) [inline modelforge/registry.py]: Future commit: typo

File: modelforge/registry.py [hunk 9]
```diff
@@ -6,12 +6,56 @@
...
+    :param backend: Backend which is responsible for working with model files.
+    :param log: Logger supplied by supply_backend
+    :return: None
+    """
+    try:
+        backend.reset(args.force)
+    except ExistingBackendError:
+        return 1
+
+    log.info("Deleting content of index ...")
```
Comment (vmarkovtsev) [inline modelforge/registry.py]: Future commit: "Resetting the index"

File: modelforge/registry.py [hunk 10]
```diff
@@ -111,11 +135,16 @@ def delete_model(args: argparse.Namespace, backend: StorageBackend, log: logging
...
     :return: None
     """
-
-    meta = backend.index.remove_model(args.input)
+    try:
+        meta = backend.index.remove_model(args.input)
+        template_readme = backend.index.load_template(args.template_readme)
+        backend.index.update_readme(template_readme)
+    except ValueError:
+        return 1
```
Comment (vmarkovtsev) [inline modelforge/registry.py]: We are catching much code here, so better to split and log what has failed exactly.

File: modelforge/registry.py [hunk 11]
```diff
@@ -111,11 +135,16 @@ def delete_model(args: argparse.Namespace, backend: StorageBackend, log: logging
...
     backend.delete_model(meta)
     log.info("Updating the models index...")
-    template_readme = backend.index.load_template(args.template_readme)
-    backend.index.update_readme(template_readme)
-    backend.index.upload_index("delete", meta)
+    try:
+        backend.index.upload_index("delete", meta)
```
Comment (vmarkovtsev) [inline modelforge/registry.py]: Future commit: rename to `upload`
